### PR TITLE
benchmark html5 canvas renderer

### PR DESCRIPTION
**Pull request by madhur-tandon**

### CODE REVIEW

**Comment by madhur-tandon** [comment]
@rth @hoodmane could you guys have a look, I am not sure what the error is here. This PR attempts to benchmark the html5 canvas renderer for the blog post pending for the pyodide blog.

**Comment by rth** [comment]
Thanks @madhur-tandon , I'll look into it.

**Comment by hoodmane** [comment]
This error appears sporadically sometimes even on main so it's worth rerunning CI a couple times to make sure it's due to your changes.

**Comment by madhur-tandon** [comment]
@hoodmane Can you restart it, I don't think I have permissions

**Comment by hoodmane** [comment]
Sure, done.

**Comment by madhur-tandon** [comment]
@hoodmane apparently it still fails with the sporadic issue, maybe we need to restart the whole CI?

**Comment by hoodmane** [comment]
> maybe we need to restart the whole CI?

I tried it one more time. It is likely a consistent problem related to your changes. Restarting from beginning doesn't usually help.

**Comment by hoodmane** [comment]
Right, hopefully it replicates locally. Sometimes there is a problem only on CI though...

**Comment by ryanking13** [comment]
It looks like it's _literally_ due to the timeout. The code takes too long to run.

I was able to fix this by:

1. updating timeout

https://github.com/pyodide/pyodide/blob/e9fd6e9079a14f389a2b8c2cc1dba3a10725c606/benchmark/benchmark.py#L148

(pass `script_timeout` argument, default is set to 20 seconds)

2. or, repeating fewer times

https://github.com/pyodide/pyodide/blob/e9fd6e9079a14f389a2b8c2cc1dba3a10725c606/benchmark/benchmark.py#L100

**Comment by madhur-tandon** [comment]
@hoodmane can you please re-trigger? Since I cannot run this locally, I need to run it on CI. The failure of `build-packages` seems random since it passed on the previous commit.

**Comment by ryanking13** [comment]
> @hoodmane can you please re-trigger? Since I cannot run this locally, I need to run it on CI. The failure of `build-packages` seems random since it passed on the previous commit.

Please apply changes of the upstream main branch. We've updated statsmodels URL and checksum (#2153) due to the Github URL issue.

**Comment by madhur-tandon** [comment]
Tips to bypass
`numpy.core._exceptions._ArrayMemoryError: Unable to allocate 1.04 MiB for an array with shape (370, 370) and data type float64` ?

**Comment by ryanking13** [comment]
> Tips to bypass `numpy.core._exceptions._ArrayMemoryError: Unable to allocate 1.04 MiB for an array with shape (370, 370) and data type float64` ?
> 
> CC: @ryanking13 @hoodmane

Hmm, I'm not sure why it happens. I can reproduce it in CI, but not locally. Maybe CI has a smaller memory.

**Comment by madhur-tandon** [comment]
> Hmm, I'm not sure why it happens. I can reproduce it in CI, but not locally. Maybe CI has a smaller memory.

You mean locally but not in CI? Well, in my case, I cannot do it locally too. I face the issue described in: https://github.com/pyodide/pyodide/issues/2150
CC: @ryanking13 

PS -- If you were able to run it locally, would you be able to give me the `benchmarks.json` file that it produced? I just need the metrics for the blog post which is currently pending because of this.

**Comment by rth** [comment]
Thanks for investigating it @ryanking13 !

**Comment by madhur-tandon** [comment]
Hi @ryanking13, I have pushed the suggested changes and have also changed the `_base_fonts_url` for the renderer. I have also made some changes to the `benchmarks`. It seems the data that you gave me wasn't correct -- the error is on my part actually since I wasn't telling which backend `matplotlib` should use. Unfortunately, after cloning a fresh repo for Pyodide and then running `PYODIDE_PACKAGES="matplotlib" make` and `make benchmark`, I still face the issue described in https://github.com/pyodide/pyodide/issues/2150

Thus, I would have to request you to give me the `benchmarks.json` file again when you run this PR locally on your machine. Alternatively, we could SSH into the CI machine to take this file out but I will need permissions for that.
CC: @rth @hoodmane

**Comment by madhur-tandon** [comment]
Seems like pyodide has changed a bit since using the renderer in `benchmarks` results in
```
File "/lib/python3.9/site-packages/matplotlib/backends/html5_canvas_backend.py", line 421, in draw_text
    f.load().then(_load_font_into_web)
AttributeError: 'Future' object has no attribute 'then'
```

CC: @rth @hoodmane @ryanking13

**Comment by madhur-tandon** [comment]
From the docs. we have this:
```
When a JavaScript function returns a Promise (for example, if the function is an async function),
it is assumed that the Promise is going to do some work that uses the arguments of the function,
so it is not safe to destroy them until the Promise resolves. In this case, the proxied function returns
a Python Future instead of the original Promise. When the Promise resolves, the result is converted
to Python and the converted value is used to resolve the Future.
```

I could appreciate some guidance on how to use this new paradigm correctly.

**Comment by ryanking13** [comment]
> Seems like pyodide has changed a bit since using the renderer in `benchmarks` results in
> 
> ```
> File "/lib/python3.9/site-packages/matplotlib/backends/html5_canvas_backend.py", line 421, in draw_text
>     f.load().then(_load_font_into_web)
> AttributeError: 'Future' object has no attribute 'then'
> ```
> 
> CC: @rth @hoodmane @ryanking13

Yes... this must be a bug in our matplotlib backend. I'm having the same issue when I run this PR locally too.
Our CI is skipping matplotlib tests which [requires html5 backend](https://github.com/pyodide/pyodide/blob/950c1dcbd271f90faea0d205a7c95fcd40710a69/packages/matplotlib/test_matplotlib.py#L471-L474).

**Comment by hoodmane** [comment]
I think it should be `f.load().add_done_callback(_load_font_into_web)`.

**Comment by madhur-tandon** [comment]
@ryanking13 Skipping those tests was intentional since we are not using this backend by default. But regardless, I have reintroduced the tests. I have also incorporated @hoodmane's suggestion regarding the Futures callback stuff. Let's see (and hope) things work now!

**Comment by madhur-tandon** [comment]
Can CI be re-triggered? It seems `build-core` is a non-deterministic failure. @ryanking13 @hoodmane

**Comment by ryanking13** [comment]
> Can CI be re-triggered? It seems `build-core` is a non-deterministic failure. @ryanking13 @hoodmane

Retriggered, thanks again for keep working on this!

**Comment by madhur-tandon** [comment]
@ryanking13 It seems like I was getting the correct benchmarks, but then tests failed -- which kinda cancelled the `benchmarks` too. The major reason of failure in all these tests are related to
`selenium.common.exceptions.TimeoutException`
Can I increase the timeout on a per test level? If yes, how would that be?

**Comment by madhur-tandon** [comment]
@rth I was also thinking to use `selenium_standalone` instead of `selenium` for the fixture for the tests. Please let me know if this is a good idea.

**Comment by rth** [comment]
Yes, selenium_standalone is probably better in this case. Thanks for improving tests!

**Comment by madhur-tandon** [comment]
@rth @ryanking13 @hoodmane I could use one last help. Everything seems fine, except that tests fail because of `check_comparison` function used at the end of every test. This function, as you can see 
- waits for Fonts to be loaded
- waits for Results to be loaded

I have increased the timeout from 70 to 200 for both of them.
The failure happens because of `selenium.common.exceptions.TimeoutException` and the underlying error seems to be
```
File "/lib/python3.9/asyncio/futures.py", line 90, in __repr__
    return '<{} {}>'.format(self.__class__.__name__,
AttributeError: 'Future' object has no attribute '__class__'
```

Looking further gives us
```
Exception in callback RendererHTMLCanvas.draw_text.<locals>._load_font_into_web(<Future finis...ect FontFace]>)
at /lib/python3.9/site-packages/matplotlib/backends/html5_canvas_backend.py:391
handle: <Handle RendererHTMLCanvas.draw_text.<locals>._load_font_into_web(<Future finis...ect FontFace]>) at
/lib/python3.9/site-packages/matplotlib/backends/html5_canvas_backend.py:391>
```

This means that `def _load_font_into_web(loaded_face):` has some issues since `loaded_face` is not of correct type it seems -- it is interpreted as `<Future finis...ect FontFace]>` as per the above message.

**Comment by madhur-tandon** [comment]
Actually, everything works, but fonts are not loaded correctly. I am investigating a bit. Maybe the path where fonts are to be found is not correct OR perhaps it isn't waiting properly. Will have to check.

**Comment by madhur-tandon** [comment]
@rth @hoodmane @ryanking13 I have pushed a debugging commit which prints some values. Through this, I am kinda able to verify that fonts have been loaded. However, it seems like requesting a redraw through `self.fig.draw_idle()` doesn't work. That's why the tests will fail since the plots produced won't match the reference images present. If I generate the reference images again, they won't have the correct font. That's why I didn't upload new reference images since that would be kinda incorrect.

I tried replacing `draw_idle` by `draw` instead and see if it works and surprisingly it did.

I am gonna try one last experiment of doing `draw_idle` before `window.font_counter += 1` and see if it works. If it does, I am gonna go ahead with this, else I am gonna use `draw` instead of `draw_idle`

**Comment by madhur-tandon** [comment]
Hopefully, this PR is done now. If CI passes, I will generate the benchmarks plot from the CI for the blog tomorrow.

**Comment by madhur-tandon** [comment]
@rth @ryanking13 Could use a re-trigger of the CI to remove non-deterministic issues.

**Comment by madhur-tandon** [comment]
@ryanking13 Maybe you could run `benchmarks` locally for this PR and send me again if possible? If `selenium` didn't crash with `page crash` errors, I could have extracted the `benchmarks` from the CI itself.

**Comment by madhur-tandon** [comment]
> merge the updated tests, and maybe one or two extra matplotlib benchmarks but not all of them right?

Yes so, we currently have 8 tests (x2 because each for firefox and chrome) = 16 and 5 benchmarks (x2 because one for each renderer) = 10. We can surely merge them if we want. This PR is ready for review and merging too.

Earlier, we skipped the 8 tests for this new renderer since it is not enabled by default. But now, we enable it in the tests itself.

**Comment by madhur-tandon** [comment]
Seems like this issue is related to Chrome -- and there are some potential solutions here: https://stackoverflow.com/questions/53902507/unknown-error-session-deleted-because-of-page-crash-from-unknown-error-cannot/53970825

Which out of those should one implement? and how would one go about doing it?

**Comment by ryanking13** [comment]
> Thanks @ryanking13 I have removed my commits which were experimenting with `shm` and have also incorporated your changes. Thanks for the benchmarks too -- but I am still gonna wait for the CI once. I have also modified `get_numpy_benchmarks` so that it doesn't execute benchmarks for `matplotlib` which are covered in a different function.

Note that Circle CI has a 1-hour duration limit of running CI jobs, so a long-running task would result in a timeout. So if other benchmarks (pystone and numpy) take too much time, I recommend disabling it.

**Comment by madhur-tandon** [comment]
> I recommend disabling it.

Just to get the benchmarks, I can disable it. But then, they are enabled because that will make the PR mergeable. If this PR is merged, and if it takes longer, is there a way to increase the time limit?

**Comment by ryanking13** [comment]
> > I recommend disabling it.
> 
> Just to get the benchmarks, I can disable it. But then, they are enabled because that will make the PR mergeable. If this PR is merged, and if it takes longer, is there a way to increase the time limit?

I'm afraid there is no way to increase the Circle CI time limit. We could update the CI config and the benchmark script so that numpy benchmark and matplotlib benchmark could be run in separate CI jobs.
(That's what we are doing for building packages. Building packages take quite a long time, so we are building packages in two CI jobs)

But doing those would require some effort. So if it blocks getting the matplotlib benchmark result, I think we could do that in a separate PR, and let this PR for just getting the matplotlib benchmark, because you have already suffered enough to get the benchmark ( I really appreciate your effort on this 🙂 )

For now, let's hope the CI runs without a timeout.

**Comment by madhur-tandon** [comment]
> We could update the CI config and the benchmark script so that numpy benchmark and matplotlib benchmark could be run in separate CI jobs. 

Yes, we can perhaps tackle that next. For now, I have removed other benchmarks. Thanks @ryanking13

**Comment by madhur-tandon** [comment]
@rth @ryanking13 In any case, there are 4 results for each benchmark.
They are with the keys -- `firefox, firefox(w/ ib), chrome, chrome(w/ ib)`

I guess I will be using `firefox` and `chrome` for the graph. But what are `firefox(w/ ib)` and `chrome(w/ ib)` are for?

**Comment by madhur-tandon** [comment]
Hi everyone, the newer benchmarks are here. The new Pyodide is significantly faster than what it was in 2019.

I am attaching both plots in this PR for your reference:

compare

**Review by ryanking13 — APPROVED** [review]
Thanks @madhur-tandon, and sorry for the late reply.

Since this PR contains some critical fixes of matplotlib, I think this can be merged now.
Then, in a seperate PR, we could refactor benchmark code.

**Comment by madhur-tandon** [comment]
Yes, thanks again @ryanking13. This PR is ready to be merged and will also be needed for the demo for the blog post.
CC: @rth

**Comment by hoodmane** [comment]
@ryanking13 I think you should go ahead and merge this if you are satisfied with it. I haven't been following the conversation.

File: benchmark/benchmark.py [hunk 1]
```diff
@@ -85,24 +90,48 @@ def get_numpy_benchmarks():
...
+            continue
+        if "canvas" in str(filename) or "wasm" in str(filename):
+            content = parse_numpy_benchmark(filename)
+            content += (
+                "import numpy as np\n"
+                "_ = np.empty(())\n"
+                "setup = setup + '\\nfrom __main__ import {}'\n"
+                "from timeit import Timer\n"
+                "t = Timer(run, setup)\n"
+                "r = t.repeat(11, 20)\n"
```
Comment (rth) [inline benchmark/benchmark.py:123]: Related to the run time, for matplotlib you could try to do,
```suggestion
"r = t.repeat(repeat=3, number=1)"
```
a large `number` is for benchmarks that take ms and where the timer accuracy is going to make results less accurate. If plotting takes second, there is no reason to make several calls in a loop. Though you still need a couple of `repeat` to get statistics.

And maybe with this, we can keep all benchmarks, if it only adds a couple of minutes to the runtime.

**Review by rth — COMMENTED** [review]
Thanks for making these fixes @madhur-tandon! A few comments below otherwise LGTM.

The other thing though is that benchmarks take 7min on main and almost 1 hour on this PR. So we should reduce that runtime (cf my suggestion below) or possibly not run all of them by default (you could add a `get_matplotlib_benchmarks(full=False)` option for instance)

File: benchmark/benchmarks/canvas_custom_font.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    t = np.arange(0.0, 2.0, 0.01)
+    s = 1 + np.sin(2 * np.pi * t)
+    plt.figure()
+    plt.title("A simple Sine Curve", **f)
+    plt.plot(t, s, linewidth=1.0, marker=11)
+    plt.plot(t, t)
+    plt.grid(True)
+    plt.show()
+    plt.close("all")
+    plt.clf()
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply benchmark/benchmarks/canvas_custom_font.py:24]: Can we address this in a separate PR maybe?

[the remark this answers is not in the document]
Comment (ryanking13) [reply benchmark/benchmarks/canvas_custom_font.py:24]: > Can we address this in a separate PR maybe?

Yes, sure

File: benchmark/benchmark.py (hunk 1 above)
Comment (ryanking13) [reply benchmark/benchmark.py:123]: Let's do this in separate PR too. We could discuss which would be better:

- Running multiple benchmarks (numpy, pystone, matplotlib) in separate jobs
- Adjusting `repeat` and `number` of benchmark

**Pull request merged**